### PR TITLE
lightning: let ignore columns be compatible with tidb backend

### PR DESCRIPTION
**Pull request by glorv**
### What problem does this PR solve?
Let lightning config `mydumper.ignore-columns` be compatible with tidb backend. 

Issue Number: close #xxx 

Problem Summary:

### What is changed and how it works?
Let the kv packend generated by encodeLoop contains the column names with ignore columns filtered instead of the raw column names. 

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by kennytm — APPROVED** [review]

**Comment by kennytm** [comment]
(i think you shouldn't add the "needs-cherry-pick-5.1" label, to avoid confusing the bot)

**Comment by glorv** [comment]
> (i think you shouldn't add the "needs-cherry-pick-5.1" label, to avoid confusing the bot)

Done. Forgot v5.1 is in br repo.😅

**Comment by glorv** [comment]
ping @3pointer  PTAL
/cc @kennytm  PTAL again, as there is quite a bit changes since the last review to fix the compatibility with generated columns 😅

**Review by kennytm — APPROVED** [review]

File: br/pkg/lightning/backend/tidb/tidb.go [hunk 1]
```diff
@@ -284,22 +287,27 @@ func (enc *tidbEncoder) Encode(logger log.Logger, row []types.Datum, _ int64, co
...
 		enc.columnIdx = columnIdx
-		enc.columnCnt = columnCount
+		enc.columnCnt = columnMaxIdx + 1
 	}
 
 	// TODO: since the column count doesn't exactly reflect the real column names, we only check the upper bound currently.
 	// See: tests/generated_columns/data/gencol.various_types.0.sql this sql has no columns, so encodeLoop will fill the
 	// column permutation with default, thus enc.columnCnt > len(row).
-	if len(row) > enc.columnCnt {
+	if len(row) < enc.columnCnt {
```
Comment (gozssky) [inline br/pkg/lightning/backend/tidb/tidb.go:310]: Could you please explain why `len(row)` should be greater than or equal to `columnCnt`(which is actual `columenMaxIdx + 1`)?

Suppose a table has 5 columns, namely `a`, `b`, `c`, `d`, `e`. We have a SQL statement in a source file like `insert into t(a, d, c) values(xxx, xxx, xxx)`. Then the `columnPermutation` should be `(0, -1, 2, 1, -1, -1)`, `columnCnt = columnMaxIdx + 1 = 4`.  `len(row) ` is 3 which is smaller than `columnCnt`, so in this scenario, data cannot be imported correctly.

Is my understanding correct？

Comment (glorv) [reply br/pkg/lightning/backend/tidb/tidb.go:310]: @gozssky In this case, the columnMaxIdx should be 2, so columnCnt = 3.

This check is used to handle there are ignored columns. Giving a source file `insert into t(a, d, c, f) values(xxx, xxx, xxx, xx)`, and ignore field `f` in lightning's config, this is still a valid source file. Then len(row) (4) will bigger than columnCnt (3).

BTW, there may be case that row count is smaller than column count if user manually create table schema has extra column with default value, e.g. `updated timestamp default current_timestamp`, then current lightning will result in an error.

Comment (gozssky) [reply br/pkg/lightning/backend/tidb/tidb.go:310]: Oh, I mistook `i` for `idx` 😂.

File: br/tests/lightning_duplicate_detection/run.sh [hunk 2]
```diff
@@ -22,7 +22,7 @@ LOG_FILE1="$TEST_DIR/lightning-duplicate-detection1.log"
 LOG_FILE2="$TEST_DIR/lightning-duplicate-detection2.log"
 
 # let lightning run a bit slow to avoid some table in the first lightning finish too fast.
-export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/lightning/restore/SlowDownImport=sleep(50)"
+export GO_FAILPOINTS="github.com/pingcap/tidb/br/pkg/lightning/restore/SlowDownImport=sleep(250)"
```
Comment (gozssky) [inline br/tests/lightning_duplicate_detection/run.sh:25]: The test `lightning_distributed_import` also need to inscrease the timeout.

**Review by gozssky — APPROVED** [review]
rest LGTM

**Pull request merged**